cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,382 @@
+ * need to first be initialized.
+ *
+ * @tparam vertex_t
+ * @tparam edge_t
+ * @tparam mask_t
+ */
+template <typename vertex_t, typename edge_t, typename mask_t = std::uint32_t>
+struct graph_mask_t {
+ public:
+  using vertex_type = vertex_t;
+  using edge_type   = edge_t;
+  using mask_type   = mask_t;
+  using size_type   = std::size_t;
+
+  ~graph_mask_t()                       = default;
+  graph_mask_t(graph_mask_t&&) noexcept = default;
+
+  graph_mask_t() = delete;
+
+  explicit graph_mask_t(raft::handle_t const& handle,
+                        vertex_t n_vertices,
+                        edge_t n_edges,
+                        bool complement = false)
+    : handle_(handle),
+      n_vertices_(n_vertices),
+      n_edges_(n_edges),
+      edges_(0, handle.get_stream()),
+      vertices_(0, handle.get_stream()),
+      complement_(complement)
+  {
+  }
+
+  explicit graph_mask_t(graph_mask_t const& other)
+    : handle_(other.handle_),
+      n_vertices_(other.n_vertices_),
+      n_edges_(other.n_edges_),
+      edges_(other.edges_, other.handle_.get_stream()),
+      vertices_(other.vertices_, other.handle_.get_stream()),
+      complement_(other.complement_)
+  {
+  }
+
+  graph_mask_t& operator=(graph_mask_t&&) noexcept = default;
+  graph_mask_t& operator=(graph_mask_t const& other) = default;
+
+  /**
+   * Determines whether the 1 bit in a vertex or edge position
+   * represents an inclusive mask or exclusive mask. Default is
+   * an inclusive mask (e.g. 1 bit means the corresponding vertex
+   * or edge should be included in computations).
+   * @return
+   */
+  bool is_complemented() const { return complement_; }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with an edge mask.
+   * @return
+   */
+  bool has_edge_mask() const { return get_edge_mask().has_value(); }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with a vertex mask.
+   * @return
+   */
+  bool has_vertex_mask() const { return get_vertex_mask().has_value(); }
+
+  /**
+   * Returns the edge mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_edge_mask() const
+  {
+    return edges_.size() > 0 ? std::make_optional<mask_t const*>(edges_.data()) : std::nullopt;
+  }
+
+  /**
+   * Retuns the vertex mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_vertex_mask() const
+  {
+    return vertices_.size() > 0 ? std::make_optional<mask_t const*>(vertices_.data())
+                                : std::nullopt;
+  }
+
+  vertex_t get_n_vertices() { return n_vertices_; }
+
+  edge_t get_n_edges() { return n_edges_; }
+
+  edge_t get_edge_mask_size() const { return n_edges_ >> log_bits<mask_t>(); }
+
+  vertex_t get_vertex_mask_size() const { return n_vertices_ >> log_bits<mask_t>(); }
+
+  void initialize_edge_mask(bool init = 0)
```

Comment:
I'm drawing a very clear distinction here. To allocate a mask just means memory has been allocated (and cleared). The initialize actually sets the values to the user's choosing. This doesn't have to do with complement at all- for example in testing, I found I wanted to initialize a mask with all 1's so that I could unset just a few of the bits. I made that available in the initialization so that we can keep that logic in the mask itself.